nipype/interfaces/utility/tests/test_wrappers.py

```diff
@@ -50,6 +52,74 @@ def increment_array(in_array):
     wf.run()
 
 
+def test_function_inputs():
+    def f(a, b=2):
+        return a + b
+
+    def g(a, b=3, **kwargs):
+        c = kwargs.pop('c', 0)
+        return a + b + c
+
+    fun1 = utility.Function(function=f)
+    fun2 = utility.Function(function=f, input_names=['a', 'b'])
+    fun3 = utility.Function(function=f, input_names=['a'])
+
+    fun1.inputs.a = 1
+    fun2.inputs.a = 1
+    fun3.inputs.a = 1
+
+    assert fun1.run().outputs.out == 3
+    assert fun2.run().outputs.out == 3
+    assert fun3.run().outputs.out == 3
+
+    fun1.inputs.b = 1
+    fun2.inputs.b = 1
+
+    assert fun1.run().outputs.out == 2
+    assert fun2.run().outputs.out == 2
+
+    with pytest.raises(traits.TraitError):
+        fun3.inputs.b = 1
+
+    # Non-arguments always rejected
+    with pytest.raises(traits.TraitError):
+        fun1.inputs.c = 1
+    with pytest.raises(traits.TraitError):
+        fun2.inputs.c = 1
+    with pytest.raises(traits.TraitError):
+        fun3.inputs.c = 1
+
+    fun_g = utility.Function(function=g, input_names=['a'])
+
+    fun_g.inputs.a = 1
+    assert fun_g.run().outputs.out == 4
+    with pytest.raises(traits.TraitError):
+        fun_g.inputs.b = 1
+    fun_g.inputs.d = 3
+    assert fun_g.run().outputs.out == 4
+    fun_g.inputs.c = 3
+    assert fun_g.run().outputs.out == 7
+
+
+def test_function_string_reset():
+    fstr = "def f(a, b):\n\treturn a + b"
+    gstr = "def g(a, b, c):\n\treturn a + b - c"
+
+    fun1 = utility.Function(function=fstr)
+
+    fun2 = utility.Function()
+    fun2.inputs.function_str = fstr
+
+    for fun in (fun1, fun2):
+        assert fun._input_names == ['a', 'b']
+        fun.inputs.a = 5
+        fun.inputs.b = 3
+        assert fun.run().outputs.out == 8
+
+        fun.inputs.function_str = gstr
```

Comment:
does this not raise an error?

Reply:
Logs a warning.

Reply:
i would just raise error just like the other immutable traits. if we have decided that this is not possible, we should throw an exception.

Reply:
Well, I implemented those just by making `traits.Enum(value)`. For this I went ahead and made a full-blown `SetOnce` trait. You can pass it an inner trait, so `function_str = SetOnce(traits.Str)`.

Added tests to make sure you can pass the same value multiple times without causing an error.